src/main/scala/org/allenai/pipeline/Workflow.scala

```diff
@@ -180,21 +183,21 @@ object Workflow {
   def forPipeline(steps: Iterable[(String, PipelineStep)], targets: Iterable[String], title: String, pipescripts: Option[PipescriptSources]): Workflow = {
     val idToName = steps.map { case (k, v) => (v.stepInfo.signature.id, k) }.toMap
     val nameToStep = steps.toMap
-    def findNodes(s: PipelineStep): Iterable[PipelineStep] =
-      Seq(s) ++ s.stepInfo.dependencies.flatMap {
-        case (name, step) =>
-          findNodes(step)
+    val allSteps = scala.collection.mutable.Set.empty[PipelineStep]
+    def findSteps(s: PipelineStep): Unit = {
+      if (!allSteps.contains(s)) {
+        allSteps += s
+        s.stepInfo.dependencies.values.foreach(findSteps)
       }
-
-    val nodeList = for {
-      name <- targets
-      step = nameToStep(name)
-      childStep <- findNodes(step)
-    } yield {
-      val id = childStep.stepInfo.signature.id
-      val childName = idToName.getOrElse(id, childStep.stepInfo.className)
-      (id, Node(childName, childStep))
     }
+    for (stepName <- targets) findSteps(nameToStep(stepName))
```

Comment:
I don't think it maps cleanly to a fold because it's recursive on the nodes of the graph.